File: backend/app/models/overview.py
from __future__ import annotations

from app.models.common import ApiModel
from app.models.tracked_application import TrackedApplicationOutput


class OverviewStat(ApiModel):
    applications_sent: int
    avg_match: int
    interviews: int
    offers: int


class OverviewTimePoint(ApiModel):
    label: str
    count: int


class OverviewOutcome(ApiModel):
    status: str
    count: int


class TopMatch(ApiModel):
    job_posting_id: str
    role: str
    company: str
    location: str | None
    match_score: int


class OverviewResponse(ApiModel):
    stats: OverviewStat
    over_time: list[OverviewTimePoint]
    outcomes: list[OverviewOutcome]
    top_matches: list[TopMatch]
    recently_applied: list[TrackedApplicationOutput]

File: webapp/src/schemas/overview.ts
import { z } from 'zod';
import { trackedApplicationSchema } from './tracked-application';

export const overviewSchema = z.object({
  stats: z.object({
    applications_sent: z.number(),
    avg_match: z.number(),
    interviews: z.number(),
    offers: z.number(),
  }),
  over_time: z.array(z.object({ label: z.string(), count: z.number() })),
  outcomes: z.array(z.object({ status: z.string(), count: z.number() })),
  top_matches: z.array(z.object({
    job_posting_id: z.string(),
    role: z.string(),
    company: z.string(),
    location: z.string().nullable(),
    match_score: z.number(),
  })),
  recently_applied: z.array(trackedApplicationSchema),
});

export type Overview = z.infer<typeof overviewSchema>;
